src/coreclr/src/jit/lsra.cpp

```diff
@@ -2223,16 +2223,33 @@ BasicBlock* LinearScan::findPredBlockForLiveIn(BasicBlock* block,
                                                BasicBlock* prevBlock DEBUGARG(bool* pPredBlockIsAllocated))
 {
     BasicBlock* predBlock = nullptr;
+    assert(*pPredBlockIsAllocated == false);
 
     // Blocks with exception flow on entry use no predecessor blocks, as all incoming vars
     // are on the stack.
     if (blockInfo[block->bbNum].hasEHBoundaryIn)
+    {
+        JITDUMP("\n\nIncoming EH boundary; ");
+        return nullptr;
+    }
+
+    if (block == compiler->fgFirstBB)
     {
         return nullptr;
     }
 
+    if (block->bbPreds == nullptr)
+    {
+        // We may have throw blocks with no predecessors, or unreachable blocks.
```

Comment:
Could you please give me an example of such blocks? Are they arm-specific blocks that we can't delete because we use them to continue after a catch block?

Reply:
There seem to be at least a couple of cases. We can eliminate control flow where we are for one reason or other unable to delete the associated blocks. Also, some EH blocks appear to have no predecessor.

Reply:
That part still is not clear for me. Does it mean that if `findPredBlockForLiveIn` returns nullptr then all incoming live-in vars will be living on the stack? I do not see where it happens and why it should happen for blocks that are not `hasEHBoundaryIn`.

> We can eliminate control flow where we are for one reason or other unable to delete the associated blocks.

In such cases, I believe we delete all instructions from these block and zero-initialize live-in and live-out sets, don't we?


> Also, some EH blocks appear to have no predecessor.

For them, we should probably return after in `if (blockInfo[block->bbNum].hasEHBoundaryIn) return nullptr'` block.

Reply:
> Also, some EH blocks appear to have no predecessor.

AFAIK all handlers do not have predecessors, are there exceptions?

Reply:
The comment is not entirely correct. If we don't have a predecessor, then we will add DummyDefs to avoid the default behavior that everything's on the stack (except for the case of the first block or blocks with incoming EH flow). However, this adds unnecessary compile-time cost and may actually result in more mismatches at the boundaries. So we're better off to simply use the variable locations from the previous block.

I have encountered non-empty non-EH blocks with no predecessor that resulted from optimizations that deleted the branch to it, but not all the instructions - I believe it is the case that they are marked in such a way that they can't be deleted.

> AFAIK all handlers do not have predecessors, are there exceptions?

A finally block, for example, is considered to have an incoming EH boundary (all vars must be on the stack), but it will have predecessors.